### PR TITLE
Fix "connection already closed" issue with dbt 1.0 and dbt_utils

### LINKED ISSUE #2297: dbt utils connection errors on 0.9.1
After upgrading to 0.9.1 to solve https://github.com/sqlfluff/sqlfluff/issues/2113 I'm seeing a new error with dbt models using dbt-utils macros.

During linting, error `connection already closed` occurs. At least one other person is experiencing this (see above issue).

<details>
<summary>Full error output</summary>
<br />

```shell
=== [dbt templater] Sorting Nodes...
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
Traceback (most recent call last):
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/adapters/postgres/connections.py", line 59, in exception_handler
    yield
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/adapters/sql/connections.py", line 73, in add_query
    cursor = connection.handle.cursor()
psycopg2.InterfaceError: connection already closed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/bin/sqlfluff", line 8, in <module>
    sys.exit(cli())
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/sqlfluff/cli/commands.py", line 497, in lint
    result = lnt.lint_paths(
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 974, in lint_paths
    self.lint_path(
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 926, in lint_path
    for i, linted_file in enumerate(runner.run(fnames, fix), start=1):
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/sqlfluff/core/linter/runner.py", line 101, in run
    for fname, partial in self.iter_partials(fnames, fix=fix):
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/sqlfluff/core/linter/runner.py", line 54, in iter_partials
    for fname, rendered in self.iter_rendered(fnames):
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/sqlfluff/core/linter/runner.py", line 43, in iter_rendered
    yield fname, self.linter.render_file(fname, self.config)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 652, in render_file
    return self.render_string(raw_file, fname, config, encoding)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 629, in render_string
    templated_file, templater_violations = self.templater.process(
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/sqlfluff_templater_dbt/templater.py", line 331, in process
    processed_result = self._unsafe_process(fname_absolute_path, in_str, config)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/sqlfluff_templater_dbt/templater.py", line 457, in _unsafe_process
    node = self.dbt_compiler.compile_node(
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/compilation.py", line 550, in compile_node
    node = self._compile_node(node, manifest, extra_context)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/compilation.py", line 388, in _compile_node
    compiled_node.compiled_sql = jinja.get_rendered(
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/clients/jinja.py", line 597, in get_rendered
    return render_template(template, ctx, node)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/clients/jinja.py", line 548, in render_template
    return template.render(ctx)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 1, in top-level template code
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/clients/jinja.py", line 332, in __call__
    return self.call_macro(*args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/clients/jinja.py", line 259, in call_macro
    return macro(*args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "<template>", line 2, in template
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/clients/jinja.py", line 332, in __call__
    return self.call_macro(*args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/clients/jinja.py", line 259, in call_macro
    return macro(*args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "<template>", line 5, in template
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/clients/jinja.py", line 332, in __call__
    return self.call_macro(*args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/clients/jinja.py", line 259, in call_macro
    return macro(*args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "<template>", line 10, in template
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/adapters/base/impl.py", line 225, in execute
    return self.connections.execute(
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/adapters/sql/connections.py", line 127, in execute
    _, cursor = self.add_query(sql, auto_begin)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/adapters/sql/connections.py", line 83, in add_query
    return connection, cursor
  File "/Users/noah/.pyenv/versions/3.8.12/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/noah/.local/share/virtualenvs/dbt-transform-HZqNz6yE/lib/python3.8/site-packages/dbt/adapters/postgres/connections.py", line 82, in exception_handler
    raise dbt.exceptions.RuntimeException(e) from e
dbt.exceptions.RuntimeException: Runtime Error
  connection already closed
```
</details>

## Steps to Reproduce
Create a model with something like this:
```sql
{%- set saved_var = dbt_utils.get_query_results_as_dict(
    "SELECT existing_table_column
    FROM public.some_db_table" 
    )
-%}

SELECT 1 FROM table_doesnt_matter
```

## Dialect
postgres (to Redshift db)

## Version
dbt 1.0.1
sqlfluff 0.9.1
sqlfluff-template-dbt 0.9.1

## Configuration
```toml
[tool.sqlfluff.core]
dialect = "postgres"
templater = "dbt"

[tool.sqlfluff.rules]
max_line_length = 100

[tool.sqlfluff.rules.L010] # SQL keywords
capitalisation_policy = "upper"

[tool.sqlfluff.rules.L014] # Unquoted identifiers
extended_capitalisation_policy = "lower"

[tool.sqlfluff.rules.L016] # Line length
ignore_comment_lines = true

[tool.sqlfluff.rules.L030] # Function names
capitalisation_policy = "upper"

[tool.sqlfluff.rules.L042]
forbid_subquery_in = "both"
```

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #2297 

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Review by jpy-git — COMMENTED** [review]
@barrywhart Nice one, that hide whitespace pr trick you mentioned the other day came in very handy here too 😄 

Do you think it's worth adding a test case using one of the `dbt_utils` macros?

**Comment by barrywhart** [comment]
@jpy-git: Yes, I just updated a test and pushed it. I was holding off pushing that until I saw that the pre-existing dbt tests seemed to be working in CI. (Seems I still can't successfully run some of them locally for some reason?!)

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py [hunk 1]
```diff
@@ -547,10 +541,19 @@ def make_template(in_str):
             [],
         )
 
-    def _slice_template(self, in_str: str) -> List[RawFileSlice]:
```
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:550]: Unrelated cleanup: This function has not been used for the past few months.

**Comment by barrywhart** [comment]
@jpy-git: I think this is ready for final review. I've been fussing a bit with getting the pre and post-templated files to agree, but it's just a missing or extra newline; I don't expect any more "real" changes.

**Review by jpy-git — APPROVED** [review]
@barrywhart Awesome work! LGTM 🚀 ✨

**Comment by barrywhart** [comment]
Oops, my last cleanup introduced some coverage gaps. I'll open a follow-on PR to add `#pragma: no cover` to those.

**Comment by jpy-git** [comment]
Oh whoops I merged after the tests but before the coverage comment 😅

**Comment by tunetheweb** [comment]
> Oh whoops I merged after the tests but before the coverage comment 😅

image

@alanmcruickshank

**Comment by NiallRees** [comment]
Thank you @barrywhart!
